File: Matlab Code/Classes/PriceModel.m
% A Pricemodel can represent costs or income. We need a name for the item, so we can put
% it on a balance sheet, then we need a unit for the price to refer to, and
% then we need the price itself, which will be represented by a Trend.
% So we have a name, a denominatorUnit, and a trend. The unit will always
% be a Money type unit, usually Dollars, but possibly cents, or possibly
% foreign currency.
%
% PriceModels should be defined for each income and cost for a given crop.
% The growthModelDelegate will be responsible for the income and costs from
% it's events. The growthModelDelegate constructor should define a list of
% productPriceModels and costPriceModels, such that the name and the Units are defined.
% The trend can be initialised to an empty trend.
% A growthModelDelegate should not be ready until all the priceModels have
% non-empty trends. It's ok if they are specifically the Zero trend.
% 
% We also add in a notes field. This can be set by the developer to send
% messages to the user, but the user can change this at will. It will stay
% with the priceModel though and be shown when the priceModel trend is
% being set. It is saved as a list of paragraphs, each paragraph saved as a
% string in a cell array.
classdef PriceModel
    
    properties       
        name = '';
        trend = Trend.empty(1, 0);
        unit = Unit.empty(1, 0);
        denominatorUnit = Unit.empty(1, 0);
        notesParagraphs = {};
        denominatorUnitIsCurrent = true;
    end
    
    properties(SetAccess = private)
        % Says whether we can change the denominatorUnit in the crop wizard
        % costs page.
        % All extraCostPriceModels should set this to false when
        % constructing the PriceModel.
        allowCostUnitChanges = true; 
    end
    
    properties(Dependent)
         
        % Gives the name in HTML and changes the colour to red if the trend
        % is invalid, or green if it's ready to go.
        markedUpName
        
    end
    
    methods
        % Constuctor for the PriceModel. Doesn't construct the Trend.
        function pm = PriceModel(name, unit, denominatorUnit, allowCostUnitChanges)
           if nargin >= 3
                pm.name = name; 
                pm.unit = unit;
                pm.denominatorUnit = denominatorUnit;
                if nargin == 4
                    pm.allowCostUnitChanges = allowCostUnitChanges;
                end
           end
        end
        
        % Get the markedUpName. Simply colours the name green if trend is
        % valid or red if it isn't.
        function markedUpName = get.markedUpName(obj)
        
            if Trend.isValid(obj.trend) && PriceModel.isValid(obj)
                markedUpName = ['<HTML><FONT color="green">', obj.name, '</FONT></HTML>'];
            else
                if ~obj.denominatorUnitIsCurrent
                    markedUpName = ['<HTML><FONT color="red">', obj.name, ' [Units No Longer Current]</FONT></HTML>'];
                else
                    markedUpName = ['<HTML><FONT color="red">', obj.name, '</FONT></HTML>'];
                end
            end            
        end
        
        function obj = markDenominatorUnitValidity(obj, TF)
            if islogical(TF)
               obj.denominatorUnitIsCurrent = TF; 
            end
        end
    end
    
    methods(Static)
       
        function valid = isValid(pm) 
            valid = isa(pm, 'PriceModel');
            valid = valid && pm.denominatorUnitIsCurrent;
            valid = all([valid ischar(pm.name), Unit.isValid(pm.unit), Unit.isValid(pm.denominatorUnit)]);
            valid = all([valid ~isempty(pm.name), ~isempty(pm.unit), ~isempty(pm.denominatorUnit)]); 
        end
        
        function ready = isReady(pm)
           ready = PriceModel.isValid(pm);
           if ready
              ready = all([Trend.isValid(pm.trend)]); 
           end
           if ready
              ready = all([~isempty(pm.Trend)]); 
           end
        end
        
        % Returns true if the name and units of a and b match.
        % Requires a and b to be priceModels.
        function match = definitionMatches(obja, objb)
            match = isa(obja, 'PriceModel') && isa(objb, 'PriceModel');
            if match 
               match = all([strcmp(obja.name, objb.name), ...
                   strcmp(obja.unit.specificName, objb.unit.specificName), strcmp(obja.denominatorUnit.specificName, objb.denominatorUnit.specificName)]);%, ...
                   %strcmp(obja.unit.speciesName, objb.unit.speciesName), strcmp(obja.denominatorUnit.speciesName, objb.denominatorUnit.speciesName)]);
            end
        end

    end
    
end

File: imagine/core/price_model.py
import imagine.core.global_helpers as global_helpers
from .amount import Amount
import copy
from .rate import Rate
from .trend import Trend
from .unit import Unit
from imagine.util.absorb_fields import absorb_fields


class PriceModel:
    def __init__(self, name: str = "", unit: Unit = Unit(),
                 denominator_unit: Unit = Unit(), allow_cost_unit_changes: bool = True):
        if isinstance(name, dict):
            # In case we pass a dictionary as first arg, load fields from that.
            d = name
            self.load_price_model(d)
        else:
            self.event_name = name
            self.name = self.event_name + " cost"
            self.unit = unit
            self.trend = Trend()
            self.denominator_unit = denominator_unit
            self.allow_cost_unit_changes = allow_cost_unit_changes
            if not unit.measurable:
                unit = copy.deepcopy(unit)
                unit.measurable = self.name
            self.rate = Rate(0, unit, denominator_unit)

        self.notes_paragraphs = []
        self.denominator_unit_is_current = True

    # @property
    # def marked_up_name(self) -> str:
    #     if Trend.is_valid(self.trend) and self.is_valid():
    #         return f'<HTML><FONT color="green">{self.name}</FONT></HTML>'
    #     elif not self.denominator_unit_is_current:
    #         return f'<HTML><FONT color="red">{self.name} [Units No Longer Current]</FONT></HTML>'
    #     else:
    #         return f'<HTML><FONT color="red">{self.name}</FONT></HTML>'

    def mark_denominator_unit_validity(self, is_current: bool):
        if isinstance(is_current, bool):
            self.denominator_unit_is_current = is_current

    def is_valid(self) -> bool:
        return isinstance(self.name, str) and Unit.is_valid(self.unit) and Unit.is_valid(self.denominator_unit)

    @staticmethod
    def is_ready(pm) -> bool:
        return PriceModel.is_valid(pm) and Trend.is_valid(pm.trend)

    # @staticmethod
    # def definition_matches(a, b) -> bool:
    #     if isinstance(a, PriceModel) and isinstance(b, PriceModel):
    #         return (a.name == b.name and a.unit.specific_name == b.unit.specific_name and
    #                 a.denominator_unit.specific_name == b.denominator_unit.specific_name)
    #     return False

    def load_price_model(self, d):
        self.allow_cost_unit_changes = True
        simple_fields = ["name", "event_name", "unit", "denominator_unit", "allow_cost_unit_changes"]
        absorb_fields(self, d, simple_fields)

        if "units" in d:
            r = d["units"]
            if isinstance(r, str):
                r = getattr(global_helpers, r, None)
            if not isinstance(r, Rate):
                if isinstance(r, Unit):
                    r = Rate(0, r, global_helpers._unit)
                else:
                    raise ValueError(f"Unable to parse rate from price model definition: {d['units']}")
            self.rate = r
            self.unit = r.unit
            self.denominator_unit = r.denominator_unit
        if not self.rate.unit.measurable:
            self.rate = copy.deepcopy(self.rate)
            self.rate.unit.measurable = self.name

        if "trend" in d:
            t = d['trend']
            if isinstance(t, dict):
                t = Trend(t)
            if isinstance(t, Trend):
                self.trend = t
            else:
                raise ValueError(f"Can't parse trend in price model configuration: {t}")

